orchagent/fdborch.cpp

```diff
@@ -617,98 +836,324 @@ void FdbOrch::updateVlanMember(const VlanMemberUpdate& update)
-    fdb_entry.bv_id = entry.bv_id;
+    if (!m_portsOrch->getPort(entry.bv_id, vlan))
+    {
+        SWSS_LOG_NOTICE("addFdbEntry: Failed to locate vlan port from bv_id 0x%lx", entry.bv_id);
+        return false;
+    }
 
-    Port port;
     /* Retry until port is created */
-    if (!m_portsOrch->getPort(entry.port_name, port))
+    if (!m_portsOrch->getPort(port_name, port) || (port.m_bridge_port_id == SAI_NULL_OBJECT_ID))
     {
-        SWSS_LOG_DEBUG("Saving a fdb entry until port %s becomes active",
-                        entry.port_name.c_str());
-        saved_fdb_entries[entry.port_name].push_back({entry, type});
+        SWSS_LOG_INFO("Saving a fdb entry until port %s becomes active", port_name.c_str());
+        saved_fdb_entries[port_name].push_back({entry.mac,
+                vlan.m_vlan_info.vlan_id, fdbData});
+        return true;
+    }
 
+    /* Retry until port is member of vlan*/
+    if (vlan.m_members.find(port_name) == vlan.m_members.end())
+    {
+        SWSS_LOG_INFO("Saving a fdb entry until port %s becomes vlan %s member", port_name.c_str(), vlan.m_alias.c_str());
+        saved_fdb_entries[port_name].push_back({entry.mac,
+                vlan.m_vlan_info.vlan_id, fdbData});
         return true;
     }
 
-    /* Retry until port is added to the VLAN */
-    if (!port.m_bridge_port_id)
+    sai_status_t status;
+    sai_fdb_entry_t fdb_entry;
+    fdb_entry.switch_id = gSwitchId;
+    memcpy(fdb_entry.mac_address, entry.mac.getMac(), sizeof(sai_mac_t));
+    fdb_entry.bv_id = entry.bv_id;
+
+    Port oldPort;
+    string oldType;
+    FdbOrigin oldOrigin = FDB_ORIGIN_INVALID ;
+    bool macUpdate = false;
+    auto it = m_entries.find(entry);
+    if(it != m_entries.end())
     {
-        SWSS_LOG_DEBUG("Saving a fdb entry until port %s has got a bridge port ID",
-                        entry.port_name.c_str());
-        saved_fdb_entries[entry.port_name].push_back({entry, type});
+        /* get existing port and type */
+        oldType = it->second.type;
+        oldOrigin = it->second.origin;
 
-        return true;
+        if (!m_portsOrch->getPortByBridgePortId(it->second.bridge_port_id, oldPort))
+        {
+            SWSS_LOG_ERROR("Existing port 0x%lx details not found", it->second.bridge_port_id);
+            return false;
+        }
+
+        if((oldOrigin == fdbData.origin) && (oldType == fdbData.type) && (port.m_bridge_port_id == it->second.bridge_port_id))
+        {
+            /* Duplicate Mac */
+            SWSS_LOG_NOTICE("FdbOrch: mac=%s %s port=%s type=%s origin=%d is duplicate", entry.mac.to_string().c_str(),
+                    vlan.m_alias.c_str(), port_name.c_str(),
+                    fdbData.type.c_str(), fdbData.origin);
+            return true;
+        }
+        else if(fdbData.origin != oldOrigin)
+        {
+            /* Mac origin has changed */
+            if((oldType == "static") && (oldOrigin == FDB_ORIGIN_PROVISIONED))
```

Comment:
I think this function has become pretty big. Can we have the vxlan related changes handled in a separate function?

Reply:
Most of this function is actually VxLan related, it doesn't help much separating that out.